pennylane/gradients/param_shift_hessian.py

```diff
@@ -0,0 +1,256 @@
+            of the returned list of tapes will match the length of ``shifts``.
+    """
+    params = list(tape.get_parameters())
+    tapes = []
+
+    for shift in shifts:
+        new_params = params.copy()
+        shifted_tape = tape.copy(copy_operations=True)
+
+        for i, s in enumerate(shift):
+            new_params[idx[i]] = new_params[idx[i]] + qml.math.convert_like(s, new_params[idx[i]])
+
+        shifted_tape.set_parameters(new_params)
+        tapes.append(shifted_tape)
+
+    return tapes
+
+
+def compute_hessian_tapes(tape, diff_methods, f0=None):
+    r"""Generate the Hessian tapes that are used in the computation of the second derivative of a
+    quantum tape, using analytical parameter-shift rules to do so exactly. Also define a
+    post-processing function to combine the results of evaluating the gradient tapes.
+
+    Args:
+        tape (.QuantumTape): input quantum tape
+        diff_methods (list[string]): The gradient method to use for each trainable parameter.
+            Can be "A" or "0", where "A" is the analytical parameter shift rule and "0" indicates
+            a 0 gradient (that is the parameter does not affect the tapes output).
+        f0 (tensor_like[float] or None): Output of the evaluated input tape. If provided,
+            and the gradient recipe contains an unshifted term, this value is used,
+            saving a quantum evaluation.
+
+    Returns:
+        tuple[list[QuantumTape], function]: A tuple containing a list of generated tapes, in
+            addition to a post-processing function to be applied to the results of the evaluated
+            tapes.
+    """
+    h_dim = tape.num_params
+
+    gradient_tapes = []
+    gradient_coeffs = []
+    unshifted_coeffs = {}
+    shapes = []
+
+    # The Hessian for a 2-term parameter-shift rule can be expressed via the following recipes.
+    # Off-diagonal elements of the Hessian require shifts to two different parameter indices.
+    # A recipe can thus be expressed via the tape patterns:
+    #       [[coeff, dummy], [mult, dummy], [shift1, shift2]]    (dummy values for ndarray creation)
+    # Each corresponding to one term in the parameter-shift formula:
+    #       didj f(x) = coeff * f(mult*x + shift1*ei + shift2*ej) + ...
+    diag_recipe = [[[0.5], [1], [np.pi]], [[-0.5], [1], [0]]]
+    off_diag_recipe = [
+        [[0.25, 1], [1, 1], [np.pi / 2, np.pi / 2]],
+        [[-0.25, 1], [1, 1], [-np.pi / 2, np.pi / 2]],
+        [[-0.25, 1], [1, 1], [np.pi / 2, -np.pi / 2]],
+        [[0.25, 1], [1, 1], [-np.pi / 2, -np.pi / 2]],
+    ]
+
+    # for now assume all operations support the 2-term parameter shift rule
+    for i in range(h_dim):
+        for j in range(h_dim):
+            # optimization: only generate tapes for upper triangular matrix (j >= i)
+            # optimization: skip partial derivates that are zero
+            if j < i or diff_methods[i] == "0" or diff_methods[j] == "0":
```

Comment:
```suggestion
            if diff_methods[j] == "0":
```
In conjunction with the change of the `range` domain and the earlier check for `diff_methods[i]` above